### PR TITLE
sql: add SHOW HISTOGRAM command and `let` logictest statement

### PR DESCRIPTION
**Pull request by RaduBerinde**
#### sql: add SHOW HISTOGRAM command

`SHOW HISTOGRAM` can be used to look at a histogram, using histogram IDs
returned by `SHOW STATISTICS`.

Release note: none

#### sql: add let statement

Adding a let statement in logictests which sets a variable to the result of a
query; the variable can be used in subsequent queries. Example:

```
let $x
SELECT 1

query I
SELECT 1 + $x
----
2
```

This syntax does not conflict with placeholder usage because variable names must
start with a letter (whereas placeholders use only digits).

Adding a `SHOW HISTOGRAM` test case using this functionality.

Release note: none

### CODE REVIEW

**Comment by rytaft** [comment]
Reviewed the first few files - I'll come back and do the rest after lunch. Good stuff!

---

Reviewed 4 of 9 files at r1.
Review status: 4 of 11 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/sql/show_stats.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XJuZnoythYatBBXGO:-L-XJuZnoythYatBBXGP:b-9yer0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/show_stats.go#L163)):*
> ```Go
> 				`SELECT "tableID", "columnIDs", histogram
> 	 	 FROM system.table_statistics
> 		 WHERE "statisticID" = $1`,
> ```

The primary key of `system.table_statistics` is `(tableID, statisticID)`.  We should probably either add an index on `statisticID` or require the table name in the `SHOW HISTOGRAM` command.

---

*[pkg/sql/show_stats.go, line 183 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XKv__z_Y_fvwGkknf:-L-XKv__z_Y_fvwGkkng:bko9pm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/show_stats.go#L183)):*
> ```Go
> 				histoColIdx
> 			)
> 			if len(row) != 3 {
> ```

You can add `rowLen` or `numCols` to the end of the const list above to get rid of this magic number 3 and string "three" below.

---

*[pkg/sql/show_stats.go, line 194 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XLdl4VOYijtu4oTUZ:-L-XLdl4VOYijtu4oTU_:bjos1cd) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/show_stats.go#L194)):*
> ```Go
> 			}
> 			if row[colIDsColIdx] == tree.DNull {
> 				return nil, errors.Errorf("statistic with no column IDs")
> ```

These last two checks may not be necessary since the `tableID` and `columnIDs` columns aren't nullable.

---

*[pkg/sql/show_stats.go, line 216 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XODFFVu-vKqCHBWjT:-L-XODFFVu-vKqCHBWjU:b-rfz15y) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/show_stats.go#L216)):*
> ```Go
> 			valType := colDesc.Type.ToDatumType()
> 			if valType == nil {
> 				panic(fmt.Sprintf("unsupported column type: %s", colDesc.Type.SemanticType))
> ```

Why is this check necessary?  Looks like you're not using `valType` below.

---

*[pkg/sql/logictest/logic_test.go, line 140 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XPyjl7z5vPlaWvzlT:-L-XPyjl7z5vPlaWvzlU:b904k8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/logictest/logic_test.go#L140)):*
> ```Go
> //
> //    Options are a comma separated strings from the following:
> //      - nosort (default)
> ```

"Options are comma separated...." (remove the a)

---

*[pkg/sql/logictest/logic_test.go, line 193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XR6Sa9IoaqdruhA1z:-L-XR6Sa9IoaqdruhA2-:b-cfo5xl) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/logictest/logic_test.go#L193)):*
> ```Go
> //  - skipif <mysql/mssql/postgresql/cockroachdb>
> //    Skips the following `statement` or `query` if the argument is postgresql
> //    or cockroachdb.
> ```

Should this be "...if the argument is mysql,  mssql, postgresql, or cockroachdb"?

---

*[pkg/sql/logictest/logic_test.go, line 197 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XREruZTKsKbiFOIJu:-L-XREruZTKsKbiFOIJv:b-siob93) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/logictest/logic_test.go#L197)):*
> ```Go
> //  - onlyif <mysql/mssql/postgresql/cockroachdb>
> //    Skips the following `statement` or query if the argument is not postgresql
> //    or cockroachdb.
> ```

Ditto...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-:-L-XSWfYnXNzAj1KUX1_:b-r19dkv)*

**Comment by rytaft** [comment]
Done reviewing - just one more inline comment.

---

Reviewed 5 of 9 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/logic_test.go, line 951 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XrHMbmy3o628jKDVk:-L-XrHMbmy3o628jKDVl:b-4net5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/logictest/logic_test.go#L951)):*
> ```Go
> // isVarChar returns true if the given string is a valid name for
> // a variable (see "let"): it must start with $, followed by a letter or
> // underscore; subsequent characters can be letters, underscores, or digits.
> ```

"it must start with $, followed by a letter or underscore..."  -- I think it can only be followed by a letter, not an underscore.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-:-L-XwH_57Rs_dsijPdWF:b-c0cgnd)*

**Comment by RaduBerinde** [comment]
TFTR!

---

Review status: 8 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/show_stats.go, line 163 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XJuZnoythYatBBXGO:-L-Xw9mORyjeQ3TgYyW2:b876224) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/show_stats.go#L163)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

The primary key of `system.table_statistics` is `(tableID, statisticID)`.  We should probably either add an index on `statisticID` or require the table name in the `SHOW HISTOGRAM` command.
</blockquote></details>

I'm not too worried - even if the table has 10,000 rows, this command will still be pretty fast

---

*[pkg/sql/show_stats.go, line 183 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XKv__z_Y_fvwGkknf:-L-XwuPdVNhfU-mkzMVl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/show_stats.go#L183)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

You can add `rowLen` or `numCols` to the end of the const list above to get rid of this magic number 3 and string "three" below.
</blockquote></details>

Done.

---

*[pkg/sql/show_stats.go, line 194 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XLdl4VOYijtu4oTUZ:-L-XwyITfGHGJvgsQVVL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/show_stats.go#L194)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

These last two checks may not be necessary since the `tableID` and `columnIDs` columns aren't nullable.
</blockquote></details>

Done.

---

*[pkg/sql/show_stats.go, line 216 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XODFFVu-vKqCHBWjT:-L-Xx5egsfuV-qQk2qsA:bsdayad) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/show_stats.go#L216)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Why is this check necessary?  Looks like you're not using `valType` below.
</blockquote></details>

Ah, sorry, left over from my first iteration (where the ResultColumn actually had this type). Removed.

---

*[pkg/sql/logictest/logic_test.go, line 140 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XPyjl7z5vPlaWvzlT:-L-XxGLXna0Z6iZiHY8l:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/logictest/logic_test.go#L140)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

"Options are comma separated...." (remove the a)
</blockquote></details>

Done.

---

*[pkg/sql/logictest/logic_test.go, line 193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XR6Sa9IoaqdruhA1z:-L-XxPRM5HmaOE5TgVH_:bj6kynq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/logictest/logic_test.go#L193)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Should this be "...if the argument is mysql,  mssql, postgresql, or cockroachdb"?
</blockquote></details>

No. If you're looking at the switch where it's handled, not that in go switch cases don't fallthrough (unless you exlpicitly `fallthrough`)

---

*[pkg/sql/logictest/logic_test.go, line 951 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XrHMbmy3o628jKDVk:-L-XxYuYoGJWj-4XGh5i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/logictest/logic_test.go#L951)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

"it must start with $, followed by a letter or underscore..."  -- I think it can only be followed by a letter, not an underscore.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-:-L-XyHfiMSxS3-lYevmX:b-814g8k)*

**Comment by rytaft** [comment]
:lgtm:

just one minor question below.

---

Reviewed 1 of 3 files at r3, 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/logic_test.go, line 193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XR6Sa9IoaqdruhA1z:-L-Y3O12yd6K29TK4pRO:btxhpyu) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/logictest/logic_test.go#L193)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

No. If you're looking at the switch where it's handled, not that in go switch cases don't fallthrough (unless you exlpicitly `fallthrough`)
</blockquote></details>

Ok I see - that's a bit confusing that `mysql` and `mssql` are acceptable arguments to `skipif` but they result in a no-op.  Maybe that should be explicitly documented?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-:-L-Y425LuDYxiMOai-sl:b2ihznz)*

File: pkg/sql/logictest/logic_test.go [hunk 1]
```diff
@@ -90,20 +91,124 @@ import (
...
+//  - statement error <regexp>
+//    Runs the statement that follows and expects an
+//    error that matches the given regexp.
+//
+//  - query <typestring> <options> <label>
+//    Runs the query that follows and verifies the results (specified after the
+//    query and a ---- separator). Example:
+//      query I
+//      SELECT 1, 2
+//      ---
```
Comment (jordanlewis) [inline pkg/sql/logictest/logic_test.go]: nit: This should be `----`

File: pkg/sql/logictest/logic_test.go [hunk 2]
```diff
@@ -314,6 +423,30 @@ func (l *lineScanner) Scan() bool {
 	return ok
 }
 
+func (l *lineScanner) Text() string {
+	text := l.Scanner.Text()
+	if len(l.varMap) == 0 {
+		return text
+	}
+
+	// See if there are any $vars to replace.
```
Comment (jordanlewis) [inline pkg/sql/logictest/logic_test.go]: You could do this whole thing with a regex couldn't you? Might be a little more concise, but it's up to you.

**Review by jordanlewis — COMMENTED** [review]
The logic test portion LGTM!

**Comment by RaduBerinde** [comment]
Review status: 8 of 11 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/logic_test.go, line 193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-XR6Sa9IoaqdruhA1z:-L-Y9Rpon3c9PEeY7vHW:bjki1k7) ([raw file](https://github.com/cockroachdb/cockroach/blob/0ae46016df7c8f1d58b2cee07a4f526f4f48b5e2/pkg/sql/logictest/logic_test.go#L193)):*
<details><summary><i>Previously, rytaft wrote…</i></summary><blockquote>

Ok I see - that's a bit confusing that `mysql` and `mssql` are acceptable arguments to `skipif` but they result in a no-op.  Maybe that should be explicitly documented?
</blockquote></details>

I think that's the point of the directive, it's so you can use the same file against different databases.

---

*[pkg/sql/logictest/logic_test.go, line 128 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-Wxyk3wjPS6OgD8ekd-r4-128:-L-YBRd4NNQ4o1P9yEUA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5605971caef01e57f2dbcf8f58422df26f31c4c0/pkg/sql/logictest/logic_test.go#L128)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: This should be `----`
</blockquote></details>

Done.

---

*[pkg/sql/logictest/logic_test.go, line 432 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-Wxyk3wjPS6OgD8ekd-r4-432:-L-YBRxY5A37TnPoMROu:b-a8rfec) ([raw file](https://github.com/cockroachdb/cockroach/blob/5605971caef01e57f2dbcf8f58422df26f31c4c0/pkg/sql/logictest/logic_test.go#L432)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

You could do this whole thing with a regex couldn't you? Might be a little more concise, but it's up to you.
</blockquote></details>

Lol, I feel silly now. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439)*

**Comment by knz** [comment]
Reviewed 8 of 9 files at r1, 3 of 3 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/show_stats.go, line 142 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-YTObUKgqZBIJlHmY8:-L-YTObVn_DdtD8Y8Cug:b6c1lmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/104a77ea847eb27a141fbe7dfb9fc2f64cb4221d/pkg/sql/show_stats.go#L142)):*
> ```Go
> }
> 
> // ShowHistogram returns a SHOW HISTOGRAM statement.
> ```

Move this to a new file `show_histogram.go`.

---

*[pkg/sql/show_stats.go, line 148 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-YTS3CovZ722yG7awM:-L-YTS3CovZ722yG7awN:b-xvfj42) ([raw file](https://github.com/cockroachdb/cockroach/blob/104a77ea847eb27a141fbe7dfb9fc2f64cb4221d/pkg/sql/show_stats.go#L148)):*
> ```Go
> 	// histogram is on. However, we don't want to have a SHOW statement for which
> 	// the schema depends on its parameters.
> 	columns := sqlbase.ResultColumns{
> ```

Extract to a global definition.

---

*[pkg/sql/parser/sql.y, line 2491 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-YTi8ivJ_QkN2BMMBl:-L-YTi8jee1LSCb7SboW:b-bfbvum) ([raw file](https://github.com/cockroachdb/cockroach/blob/104a77ea847eb27a141fbe7dfb9fc2f64cb4221d/pkg/sql/parser/sql.y#L2491)):*
> ```Yacc
> // Returns the data in the histogram with the
> // given ID (as returned by SHOW STATISTICS).
> //   
> ```

extraneous spaces

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439)*

**Comment by RaduBerinde** [comment]
Review status: 7 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/show_stats.go, line 142 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-YTObUKgqZBIJlHmY8:-L-Yc8rPthEF5Wvgiy5M:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/104a77ea847eb27a141fbe7dfb9fc2f64cb4221d/pkg/sql/show_stats.go#L142)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Move this to a new file `show_histogram.go`.
</blockquote></details>

Done.

---

*[pkg/sql/show_stats.go, line 148 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-YTS3CovZ722yG7awM:-L-Yc9BsToWZGRfEiwoh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/104a77ea847eb27a141fbe7dfb9fc2f64cb4221d/pkg/sql/show_stats.go#L148)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Extract to a global definition.
</blockquote></details>

Done.

---

*[pkg/sql/parser/sql.y, line 2491 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-YTi8ivJ_QkN2BMMBl:-L-Yc9snqyybkxkpq0ua:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/104a77ea847eb27a141fbe7dfb9fc2f64cb4221d/pkg/sql/parser/sql.y#L2491)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

extraneous spaces
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439)*

**Comment by knz** [comment]
:lgtm:

---

Reviewed 2 of 5 files at r7, 2 of 2 files at r8.
Review status: 11 of 12 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/show_histogram.go, line 41 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-Yd1JjwE7ThcLG4GQ3:-L-Yd1JjwE7ThcLG4GQ4:bnsslhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5838641f13e937b976a844cb73a476d32c7e7ed/pkg/sql/show_histogram.go#L41)):*
> ```Go
> // Privileges: Any privilege on the respective table.
> func (p *planner) ShowHistogram(ctx context.Context, n *tree.ShowHistogram) (planNode, error) {
> ```

nit: stray newline

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-:-L-Yd5M3LVBQWovqhEKu:bnfp4nl)*

**Comment by RaduBerinde** [comment]
*[pkg/sql/show_histogram.go, line 41 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439#-L-Yd1JjwE7ThcLG4GQ3:-L-Yf6-wvLbkTG3EiDmL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5838641f13e937b976a844cb73a476d32c7e7ed/pkg/sql/show_histogram.go#L41)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: stray newline
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20439)*

**Pull request merged**